File: springboot-backend/src/main/java/com/capstone/project/service/TestService.java
package com.capstone.project.service;

import com.capstone.project.dto.TestandClassLearnerDTO;
import com.capstone.project.exception.ResourceNotFroundException;
import com.capstone.project.model.Post;
import com.capstone.project.model.Test;
import com.capstone.project.model.TestLearner;
import com.capstone.project.model.TestResult;
import org.springframework.stereotype.Service;

import java.text.ParseException;
import java.util.Date;
import java.util.List;
import java.util.Map;

public interface TestService {
    List<Test> getAllTest();

    Test createTest(Test test);

    Test getTestById ( int id) throws ResourceNotFroundException;

    List<Test> getAllTestByClassId(int id);
    List<Test> getTestByUser (String name) throws ResourceNotFroundException;

    Test updateTest( int id, Test test) throws ResourceNotFroundException;

    Boolean deleteTest( int id) throws ResourceNotFroundException;

    Map<String, Object> getFilterTest(String search, String author, String direction, int duration, int classid,
                                      String fromStarted, String toStarted, String fromCreated, String toCreated, Boolean isDraft, String sortBy, int page, int size) throws Exception;

    Map<String, Object> getNumAttemptTest(int testid, int classid) throws ResourceNotFroundException;

    Map<String, Object> getNumAttempt(int testid, int userid);

    List<TestandClassLearnerDTO> getFilterTestLearner(String username, Double mark, int classid, int authorid,int testid, String direction, String sortBy, int page, int size) ;

    Map<String, Object> startTest(int testId, int userId);

    Map<String, Object> endTest(List<TestResult> testResultList) throws Exception;
}

File: react-frontend/src/services/TestService.js
import axios from 'axios'

const API_BASE_URL = process.env.REACT_APP_HOST

const createTest = (test) => {
    return axios.post(API_BASE_URL + '/test', test)
}

const updateTest = (id, test) => {
    return axios.put(API_BASE_URL + '/test/' + id, test)
}

const getTestById = (id) => {
    return axios.get(API_BASE_URL + '/test/' + id)
}

const deleteTest = (id) => {
    return axios.delete(API_BASE_URL + '/test/' + id)
}

const getFilterList = (
    search,
    author,
    fromStart,
    toStart,
    fromCreated,
    toCreated,
    draft,
    direction,
    sortBy,
    duration,
    classid,
    page,
    size
) => {
    return axios.get(
        API_BASE_URL +
            '/filtertest?search' +
            search +
            '&author' +
            author +
            (fromStart ? `&fromstarted${fromStart}` : '') +
            (toStart ? `&tostarted${toStart}` : '') +
            (fromCreated ? `&fromcreated${fromCreated}` : '') +
            (toCreated ? `&tocreated${toCreated}` : '') +
            '&draft' +
            draft +
            '&direction' +
            direction +
            '&duration' +
            duration +
            '&sortby' +
            sortBy +
            '&classid' +
            classid +
            '&page' +
            page +
            '&size' +
            size
    )
}

const startTest = (userId, testId) => {
    return axios.get(
        API_BASE_URL + '/starttest?userid=' + userId + '&testid=' + testId
    )
}

const endTest = (testResultList) => {
    return axios.post(API_BASE_URL + '/endtest', testResultList)
}

const getNumAttemptTest = (testId, classId) => {
    return axios.get(
        API_BASE_URL +
            '/getattempttest?testid=' +
            testId +
            '&classid=' +
            classId
    )
}

const getTestLearner = (
    username,
    authorId,
    mark,
    direction,
    sortBy,
    page,
    size
) => {
    return axios.get(
        API_BASE_URL +
            '/gettestlearner?username' +
            username +
            '&authorid' +
            authorId +
            '&mark' +
            mark +
            '&direction' +
            direction +
            '&sortby' +
            sortBy +
            '&page' +
            page +
            '&size' +
            size
    )
}

const TestService = {
    getFilterList,
    createTest,
    updateTest,
    getTestById,
    deleteTest,
    startTest,
    endTest,
    getNumAttemptTest,
    getTestLearner,
}
export default TestService
